File: backend/app/shared/permission_modules.py
"""Module definitions for standardized RBAC (read / create / update / edit)."""

# (resource_code, display_label)
PERMISSION_MODULES: list[tuple[str, str]] = [
    ("org", "Organization"),
    ("employee", "Employees"),
    ("recruitment", "Recruitment"),
    ("leave", "Leave"),
    ("attendance", "Attendance"),
    ("performance", "Performance"),
    ("training", "Training"),
    ("onboarding", "Onboarding"),
    ("exit", "Exit & Offboarding"),
    ("project", "Projects"),
    ("task", "Tasks"),
    ("calendar", "Calendar"),
    ("document", "Documents"),
    ("communication", "Communication"),
    ("report", "Reports"),
    ("analytics", "Analytics"),
    ("inventory", "Inventory"),
    ("resource", "Resources"),
    ("finance", "Finance"),
    ("procurement", "Procurement"),
    ("manufacturing", "Manufacturing"),
    ("logistics", "Logistics"),
    ("approval", "Approvals"),
    ("workflow", "Workflows"),
    ("audit", "Audit Log"),
    ("notification", "Notifications"),
    ("search", "Search"),
    ("user", "Users"),
    ("role", "Roles"),
    ("tenant", "Tenant Settings"),
]

STANDARD_ACTIONS = ("read", "create", "update", "edit")

# Extra permissions beyond the standard four per module
SPECIAL_PERMISSIONS: list[tuple[str, str, str]] = [
    ("org", "delete", "Delete organization nodes"),
    ("org", "manage_types", "Configure hierarchy level types"),
    ("leave", "request", "Submit leave requests"),
    ("approval", "action", "Approve or reject requests"),
    ("tenant", "settings", "Manage tenant settings"),
]

File: frontend/src/config/permission-modules.ts
/** Module labels for the permission matrix UI (must match backend PERMISSION_MODULES). */

export const PERMISSION_MODULES: { code: string; label: string }[] = [
  { code: "org", label: "Organization" },
  { code: "employee", label: "Employees" },
  { code: "recruitment", label: "Recruitment" },
  { code: "leave", label: "Leave" },
  { code: "attendance", label: "Attendance" },
  { code: "performance", label: "Performance" },
  { code: "training", label: "Training" },
  { code: "onboarding", label: "Onboarding" },
  { code: "exit", label: "Exit & Offboarding" },
  { code: "project", label: "Projects" },
  { code: "task", label: "Tasks" },
  { code: "calendar", label: "Calendar" },
  { code: "document", label: "Documents" },
  { code: "communication", label: "Communication" },
  { code: "report", label: "Reports" },
  { code: "analytics", label: "Analytics" },
  { code: "inventory", label: "Inventory" },
  { code: "resource", label: "Resources" },
  { code: "finance", label: "Finance" },
  { code: "procurement", label: "Procurement" },
  { code: "manufacturing", label: "Manufacturing" },
  { code: "logistics", label: "Logistics" },
  { code: "approval", label: "Approvals" },
  { code: "workflow", label: "Workflows" },
  { code: "audit", label: "Audit Log" },
  { code: "notification", label: "Notifications" },
  { code: "search", label: "Search" },
  { code: "user", label: "Users" },
  { code: "role", label: "Roles" },
  { code: "tenant", label: "Tenant Settings" },
];

export const MODULE_ACTIONS = ["read", "create", "update", "edit"] as const;
export type ModuleAction = (typeof MODULE_ACTIONS)[number];

export const ACTION_LABELS: Record<ModuleAction, string> = {
  read: "Read",
  create: "Create",
  update: "Update",
  edit: "Edit",
};

/** Extra permissions shown below the matrix (not part of the standard four). */
export const SPECIAL_PERMISSIONS: { resource: string; action: string; label: string }[] = [
  { resource: "org", action: "delete", label: "Organization — Delete nodes" },
  { resource: "org", action: "manage_types", label: "Organization — Manage level types" },
  { resource: "leave", action: "request", label: "Leave — Submit requests" },
  { resource: "approval", action: "action", label: "Approvals — Approve / reject" },
  { resource: "tenant", action: "settings", label: "Tenant — Settings" },
];

export function permissionKey(resource: string, action: string) {
  return `${resource}:${action}`;
}
